module/zfs/zfs_ioctl.c

```diff
@@ -5534,31 +5542,64 @@ dump_bytes_cb(void *arg)
 	dbi->dbi_err = zfs_file_write(fp, buf, dbi->dbi_len, NULL);
 }
 
+typedef struct dump_bytes_arg {
+	zfs_file_t	*dba_fp;
+#ifdef USE_SEND_TASKQ
+	taskq_t		*dba_tq;
+#endif
+} dump_bytes_arg_t;
+
 static int
 dump_bytes(objset_t *os, void *buf, int len, void *arg)
 {
+	dump_bytes_arg_t *dba = (dump_bytes_arg_t *)arg;
 	dump_bytes_io_t dbi;
 
-	dbi.dbi_fp = arg;
+	dbi.dbi_fp = dba->dba_fp;
 	dbi.dbi_buf = buf;
 	dbi.dbi_len = len;
 
-#if defined(HAVE_LARGE_STACKS)
-	dump_bytes_cb(&dbi);
+#ifdef USE_SEND_TASKQ
+	taskqid_t id = taskq_dispatch(dba->dba_tq, dump_bytes_cb, &dbi,
+	    TQ_SLEEP);
+	if (id)
+		taskq_wait_id(dba->dba_tq, id);
 #else
-	/*
-	 * The vn_rdwr() call is performed in a taskq to ensure that there is
-	 * always enough stack space to write safely to the target filesystem.
-	 * The ZIO_TYPE_FREE threads are used because there can be a lot of
-	 * them and they are used in vdev_file.c for a similar purpose.
-	 */
-	spa_taskq_dispatch_sync(dmu_objset_spa(os), ZIO_TYPE_FREE,
-	    ZIO_TASKQ_ISSUE, dump_bytes_cb, &dbi, TQ_SLEEP);
-#endif /* HAVE_LARGE_STACKS */
+	dump_bytes_cb(&dbi);
+#endif
 
 	return (dbi.dbi_err);
 }
 
+static int
+dump_bytes_init(dump_bytes_arg_t *dba, int fd, dmu_send_outparams_t *out)
+{
+	zfs_file_t *fp = zfs_file_get(fd);
+	if (fp == NULL)
+		return (SET_ERROR(EBADF));
+
+	dba->dba_fp = fp;
+#ifdef USE_SEND_TASKQ
+	dba->dba_tq = taskq_create("z_send", 1, defclsyspri, 0, 0, 0);
```

Comment:
I suppose this is executed in user process context?  Would `taskq_create_proc(curproc)` look better in some `top` output?

Reply:
I am uncertain. I think it might be better, but I don't know enough FreeBSD to say for sure.

On Linux `taskq_create_proc` is just an alias for `taskq_create`, so no change there.

I played on FreeBSD a bit. With `taskq_create` the task is called `[zfskern/z_send]`, and if I don't destroy the taskq, `zfs send` would still exit. With `task_create_proc`, I couldn't find the task, andnot destroying the taskq made `zfs send` would hang.

My guess, it's just waiting for the kernel thread to finish, and I guess I just didn't find the right incantation to see the show the task.

If that's all true, then sure, we can do that. Let me know what you think; I'm happy to do what you think makes more sense here.

Reply:
Why don't we go ahead and merge this as is since I'm not hearing a strong preference for what's best for FreeBSD.  We can always change it if needed.